src/Coordination/SnapshotableHashTable.h

```diff
@@ -209,28 +227,25 @@ class SnapshotableHashTable
         clear();
     }
 
-    std::pair<typename IndexMap::LookupResult, bool> insert(const std::string & key, V value)
+    void initialize(const KeeperContextPtr & context)
     {
-        size_t hash_value = map.hash(key);
-        auto it = map.find(key, hash_value);
-
-        if (!it)
-        {
-            auto value_size = value.sizeInBytes();
-            ListElem elem{copyStringInArena(arena, key), std::move(value)};
-            elem.setVersion(current_version);
-            auto itr = list.insert(list.end(), std::move(elem));
-            bool inserted;
-            map.emplace(itr->key, it, inserted, hash_value);
-            itr->setActiveInMap();
-            chassert(inserted);
+        min_load_factor = context->getCoordinationSettings()[CoordinationSetting::hash_map_min_load_factor];
+        min_node_count_for_auto_optimize = context->getCoordinationSettings()[CoordinationSetting::min_node_count_for_auto_optimize];
+    }
 
-            it->getMapped() = itr;
-            updateDataSize(INSERT_OR_REPLACE, key.size(), value_size, 0);
-            return std::make_pair(it, true);
-        }
+    std::pair<typename IndexMap::const_iterator, bool> insert(const std::string & key, const V & value)
+    {
+        if (auto it = map.find(key); it != map.end())
+            return std::make_pair(it, false);
 
-        return std::make_pair(it, false);
+        ListElem elem{copyStringInArena(arena, key), std::move(value)};
+        elem.setVersion(current_version);
+        auto itr = list.insert(list.end(), std::move(elem));
+        auto [it, inserted] = map.emplace(itr->key, itr);
+        chassert(inserted);
+        itr->setActiveInMap();
+        updateDataSize(INSERT_OR_REPLACE, key.size(), value.sizeInBytes(), 0);
+        return std::make_pair(it, true);
```

Comment:
Shouldn't we have the `optimizeIfNeeded` here as well?

I guess it is very unlikely case that the `hash_map_min_load_factor` is satisfied, but `min_node_count_for_auto_optimize` does not, and there were no `erase` after this, only `insert`'s, but still